daemon/src/taker_cfd.rs

```diff
@@ -510,92 +391,67 @@ where
     }
 }
 
-impl<O: 'static, M: 'static, W: 'static> Actor<O, M, W>
+#[xtra_productivity]
+impl<O, M, W> Actor<O, M, W>
 where
-    Self: xtra::Handler<CfdRollOverCompleted>,
-    O: xtra::Handler<oracle::GetAnnouncement>,
-    W: xtra::Handler<wallet::TryBroadcastTransaction>
-        + xtra::Handler<wallet::Sign>
-        + xtra::Handler<wallet::BuildPartyParams>,
+    M: xtra::Handler<monitor::StartMonitoring>,
+    O: xtra::Handler<oracle::GetAnnouncement> + xtra::Handler<oracle::MonitorAttestation>,
 {
-    async fn handle_roll_over_accepted(
+    async fn handle_propose_rollover(
         &mut self,
-        order_id: OrderId,
-        oracle_event_id: BitMexPriceEventId,
+        msg: ProposeRollOver,
         ctx: &mut Context<Self>,
     ) -> Result<()> {
-        tracing::info!(%order_id, "Roll; over request got accepted");
-
-        let (sender, receiver) = mpsc::unbounded();
+        let ProposeRollOver { order_id } = msg;
 
-        if let RollOverState::Active { .. } = self.roll_over_state {
-            anyhow::bail!("Already rolling over a contract!")
-        }
+        let disconnected = self
+            .rollover_actors
+            .get_disconnected(order_id)
```

Comment:
A pattern seems to be forming here where we:

1. Get a `Disconnected` entry
2. Start a new actor
3. Add the actor to the map

Might be worth exploring an API such as:

```rust
/// Inserts a new actor under the given key if the entry is either empty or the existing address is disconnected.
pub fn insert_new(&mut self, key: K, actor_ctor: impl Fn() -> A) -> Result<(), StillConnected>;
```

That would save a couple of lines and remove the awkwardly named `disconnected` variable.

Reply:
I've attempted to implement this quickly, but in practice we run into issues with the `actor_ctor` due to the borrow checker.

Reply:
Thanks for trying, probably needs some iteration yeah!